### ISSUE TITLE
Unhandled 'error' event when JSON.parse throws in loadJSON.js

### ISSUE THREAD

**Issue by kareniel** [pelias/whosonfirst#315]
Setup:
AWS EC2 (r4 xlarge) running Ubuntu 16 with Node v7.10.1

1. I used `npm run download` with "importPlace" using wof id `101736545` to get only the city of Montreal area.
2.  It seems like the downloader failed silently on some places, like `521613455`, and the .geojson file ended up with the content of some nginx error page:

`521/613/455/521613455.geojson`
```html
<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.4.6 (Ubuntu)</center>
</body>
</html>
```

I doubled check and the data looks good on the mapzen website:

`https://whosonfirst.mapzen.com/data/521/613/455/521613455.geojson`
```json
{
  "id": 521613455,
  "type": "Feature",
  "properties": { ... },
  "bbox": [
    -73.582871,
    45.445341,    -73.582871,
    45.445341
],
  "geometry": {"coordinates":[-73.582871,45.445341],"type":"Point"}
}
```
3. I guess that's an issue in and of itself, but because it fails silently I kept going with my import.
4. I then ran `npm start` which started the import.
5. But when trying to parse `521613455.geojson`, which contains html, not json, the whole import fails:

```
2018-01-09T16:32:22.929Z - error: [whosonfirst] exception parsing JSON for id 521613455 
in file 521/613/455/521613455.geojson: 
SyntaxError: Unexpected token < in JSON at position 0
events.js:163
      throw er; // Unhandled 'error' event
      ^

SyntaxError: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at fs.readFile (/usr/local/src/pelias/whosonfirst/src/components/loadJSON.js:29:25)
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:439:3)
```

I'm going to investigate further and then report on my findings.

If there is anything you can point out that could help me in the process, I'll be super grateful :)

Related: The downloader seems to use [https://whosonfirst.mapzen.com](https://whosonfirst.mapzen.com) as a source, which, if my understanding is correct, won't be up come February 1st.

**Comment by kareniel** [comment]
Here's how I worked around this for now.

- I ran [this script](https://gist.github.com/kareniel/0a85a39457c3de659e1fd83d21914f31) to remove the files that can't be parsed as json.
- I ran the importer again and it managed to finish.
- The importer did throw an error but it didn't make the rest of the script crash.
```
 error: [whosonfirst] ENOENT: no such file or directory, open '/tmp/wof/data/521/613/455/521613455.geojson'
```

**Comment by orangejulius** [comment]
Hi @kareniel,
Thanks for reporting the issue and giving many great details. I'm pinging @thisisaaronland to comment on the possible issues with the WOF data service, and confirm whether or not something comparable to the Mapzen Places API will be available going forward.

I'm guessing it won't be. I wonder if writing our own code to download the full planet bundles (or use the SQLite files) to essentially do our own version of the filtering by area.

**Comment by kareniel** [comment]
I'm still trying to install Pelias \*_*

Something I noticed that's worth noting is that (I think) the download script for the Who's on First importer doesn't handle 502 - Bad Gateway http errors - at all.

- When it's meta data (the csv files), it doesn't write to the file and just crashes.

```
error: [download_data_filtered] 502
error: [download_data_filtered] [generateMetafiles] Error: failed to get place info
info: [download_data_filtered] Finished generating metafile for locality failed to get meta data for locality

Error: Uncaught, unspecified "error" event. (failed to get meta data for locality)
   ...stacktrace
```

- When it's a place document, it writes the nginx bad gateway html response to the file. I'm not sure if it keeps going or if it crashes. But then the importer -will- crash if it tries to parse this place document.

**Comment by orangejulius** [comment]
@kareniel I think you're right. Error handling is important here. Even something as simple as skipping the file, so that a subsequent retry would get it, would be better. Do you by chance have time to work on something like that? We would love to accept such a PR :)

**Comment by kareniel** [comment]
We use Mapzen Search at work for maintaining a database of venues consistent and for searching through Who's on First and Open Street Map places. I'm trying to replace it with a local Pelias install. I've been trying to run the script with places less granular than country, like continent. On my third try I ran out of inodes, so I wiped the hd and started again province by province, but it's pretty slow (we're I'm capped at 3 calls to Mapzen per second because we're using our account in production). 

Fixing this could mean we could download & import wof data much faster I think.
If you could guide me through such a PR, I'd be super happy to help in the best way I can. 

There's two things here, right?  Handling the error in some any way, even if it's just catching it. Then I guess we want to keep track of what failed so it can be tried again in the future.

**Comment by orangejulius** [comment]
Ouch. at 3RPS that must be extremely painful at the continent level. I'm surprised it even works. We often had trouble even at the country level, because of the number of records involved. Honestly try downloading the full bundles by removing the `importPlace` parameter in `pelias.json`.

If you really need to limit the data involved, I think that will involve some filtering in our import pipeline, which I mentioned, and we would love to integrate.

**Comment by orangejulius** [comment]
Hey @kareniel 
Just wanted to let you know that as of #324 we now have working WOF downloads again. It does not require hitting any (now dead) Mapzen APIs, and it should be able to easily handle filtering at any level, even for big countries like the USA. It also supports multiple filtering areas. So you could specify the country IDs for say, Canada, USA, and Mexico.

Please test it out and let me know how it works!

**Comment by kareniel** [comment]
@orangejulius: Hey that's good news!

On my side, I got my team to sign up for  :earth_americas: [geocode.earth](https://geocode.earth) ;)
I'd love to test it this PR out out but I don't have access to the kind of machine I was using before (r4.xlarge on aws) so I don't think it'll be feasible. I'll probably end up using Pelias for a personal project later this year but the import area will be *significantly* smaller. So, yeah :/ Thanks for the heads up!

**Comment by orangejulius** [comment]
Oh great. I guess I better process some more invites then! Stay tuned.

Feel free to test smaller extracts. The WOF data weighs in at 15GB (so you need 30GB temporarily to download and then extract), but even my relatively slow laptop can process an export of NYC or New York State in a few seconds.

**Comment by orangejulius** [comment]
This was resolved in #324. We now support WOF `sqlite` downloads as the main way to download whosonfirst data.
